Sources/Sentry/SentryBaseIntegration.m

```diff
@@ -0,0 +1,142 @@
+#import "SentryBaseIntegration.h"
+#import "SentryCrashWrapper.h"
+#import "SentryLog.h"
+#import <Foundation/Foundation.h>
+#import <SentryDependencyContainer.h>
+#import <SentryOptions+Private.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@implementation SentryBaseIntegration
+
+- (NSString *)integrationName
+{
+    return NSStringFromClass([self classForCoder]);
+}
+
+- (BOOL)installWithOptions:(SentryOptions *)options
+{
+    return [self shouldBeEnabledWithOptions:options];
+}
+
+- (void)logWithOptionName:(NSString *)optionName
+{
+    [self logWithReason:[NSString stringWithFormat:@"because %@ is disabled", optionName]];
+}
+
+- (void)logWithReason:(NSString *)reason
+{
+    [SentryLog logWithMessage:[NSString stringWithFormat:@"Not going to enable %@ %@.",
+                                        self.integrationName, reason]
+                     andLevel:kSentryLevelDebug];
+}
+
+- (BOOL)shouldBeEnabledWithOptions:(SentryOptions *)options
+{
+    SentryIntegrationOption integrationOptions = [self integrationOptions];
+
+    if ((integrationOptions & kIntegrationOptionEnableAutoSessionTracking)
+        && !options.enableAutoSessionTracking) {
+        [self logWithOptionName:@"enableAutoSessionTracking"];
+        return NO;
+    }
+
+    if ((integrationOptions & kIntegrationOptionEnableOutOfMemoryTracking)
+        && !options.enableOutOfMemoryTracking) {
+        [self logWithOptionName:@"enableOutOfMemoryTracking"];
+        return NO;
+    }
+
+    if ((integrationOptions & kIntegrationOptionEnableAutoPerformanceTracking)
+        && !options.enableAutoPerformanceTracking) {
+        [self logWithOptionName:@"enableAutoPerformanceTracking"];
+        return NO;
+    }
+
+#if SENTRY_HAS_UIKIT
+    if ((integrationOptions & kIntegrationOptionEnableUIViewControllerTracking)
+        && !options.enableUIViewControllerTracking) {
+        [self logWithOptionName:@"enableUIViewControllerTracking"];
+        return NO;
+    }
+
+    if ((integrationOptions & kIntegrationOptionAttachScreenshot) && !options.attachScreenshot) {
+        [self logWithOptionName:@"attachScreenshot"];
+        return NO;
+    }
+
+    if ((integrationOptions & kIntegrationOptionEnableUserInteractionTracing)
+        && !options.enableUserInteractionTracing) {
+        [self logWithOptionName:@"enableUserInteractionTracing"];
+        return NO;
+    }
+#endif
+
+    if (integrationOptions & kIntegrationOptionEnableAppHangTracking) {
+        if (!options.enableAppHangTracking) {
+            [self logWithOptionName:@"enableAppHangTracking"];
+            return NO;
+        }
+
+        if (options.appHangTimeoutInterval == 0) {
+            [self logWithReason:@"because appHangTimeoutInterval is 0"];
+            return NO;
+        }
+
+        if ([SentryDependencyContainer.sharedInstance.crashWrapper isBeingTraced]) {
+            [self logWithReason:@"because the debugger is attached"];
+            return NO;
+        }
```

Comment:
I think we need a different flag for this. We may want to check whether the debugger is attach in other integrations.

Reply:
I didn't want to add another option that is only used in one place, but yeah no problem! Makes it reusable right from the start and doesn't hide this logic in the base implementation.